libnetwork/sandbox.go

```diff
@@ -273,6 +276,12 @@ func (sb *Sandbox) Refresh(options ...SandboxOption) error {
 	return nil
 }
 
+// CtrTaskCreated is to be called after the container task has been created, before
+// it is started.
+func (sb *Sandbox) CtrTaskCreated() error {
+	return sb.ctrTaskCreatedDNS()
+}
```

Comment:
Please hang this stuff off the `SetKey` hook callback instead. It's the only way to hook into the container lifecycle in a way that is compatible with how we currently integrate with buildkit. Forget I ever opened #47052. I now realize it was a mistake as we can't actually make use of splitting create and start until #44385 is merged.

Reply:
Hmm - that won't work because, as described above, the `--sysctl` is applied after the SetKey excursion. So, the IPv6-ness of the container can't be determined during SetKey.

I'll dig in to it some more to understand the buildkit problem and what your PR does. Maybe the user's sysctl can be applied before the SetKey, but that's probably a rather significant change.

Reply:
Is the concern that buildkit won't call `CtrTaskCreated()`?

If so, I think it's covered, because this change makes sure it'll get the old behaviour (and so will a further change I'm working on to defer IPv6 address allocation until after detecting IPv6 support in the container). I noted in this PR's description that ...

> To avoid potential problems for anything using libnetwork but not calling the new Sandbox method after container task creation - pre-start, the hosts file is created as-before (with the assumption that IPv6 is supported). Once the true IPv6 capability is known, the hosts file is re-generated.

If the remainder of your https://github.com/moby/moby/pull/44385 guarantees that's not necessary, it'll be possible to simplify the libnetwork code a little once it's merged.

Even then, without the early SetKey (before the user's `--sysctl` is applied) - we'll need to confirm that IPv6 is enabled in the container's namespace if possible. (I'm not sure if `net.ipv6.conf.[all|default].disable_ipv6 = 1` on the host could cause problems anywhere, perhaps not.)

But, as things stand, because the `--sysctl` is applied after the SetKey prestart hook runs, we need a chance to get in there afterwards.

(A simple "docker build" works fine for me with this change - but that probably doesn't mean much, as I don't know what problem you're anticipating, so I may not be testing for it.)

Reply:
> the `--sysctl` is applied after the SetKey excursion

Huh, you're correct. RunC behaves [contrary to the OCI runtime spec](https://github.com/opencontainers/runtime-spec/blob/main/runtime.md#start:~:text=All%20of%20the%20properties%20configured%20in%20config.json%20except%20for%20process%20MUST%20be%20applied.) and applies sysctls _after_ the `create` hooks. Sigh.